h2o-algos/src/test/java/hex/CrossValidFoldAssignmentsTest.java

```diff
@@ -0,0 +1,101 @@
+package hex;
+
+import hex.genmodel.utils.DistributionFamily;
+import hex.tree.gbm.GBM;
+import hex.tree.gbm.GBMModel;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import water.DKV;
+import water.TestUtil;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.rapids.ast.prims.advmath.AstKFold;
+
+import static junit.framework.TestCase.assertNotNull;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+public class CrossValidFoldAssignmentsTest extends TestUtil {
+  @BeforeClass() public static void setup() { stall_till_cloudsize(1); }
+
+  // Check that we should not make a copy of fold assignments Vec in ModelBuilder.
+  // Fold column as a part of train dataset is not being tracked by Scope.enter() and is still available
+  // after the model is trained.
+  @Test public void checkFoldAssignmentsAreKeptWithoutMakeCopy() {
+    final int nfolds = 3;
+    Frame tfr = null;
+    Frame cvFoldAssignmentFrame = null;
+    Frame foldId = null;
+    GBMModel gbm = null;
+    try {
+      tfr = parse_test_file("smalldata/iris/iris_wheader.csv");
+      foldId = new Frame(new String[]{"foldId"}, new Vec[]{AstKFold.kfoldColumn(tfr.vec("class").makeZero(), nfolds, 543216789)});
+      tfr.add(foldId);
+      DKV.put(tfr);
+
+      GBMModel.GBMParameters parms = new GBMModel.GBMParameters();
+      parms._train = tfr._key;
+      parms._response_column = "class";
+      parms._ntrees = 1;
+      parms._max_depth = 1;
+      parms._fold_column = "foldId";
+      parms._distribution = DistributionFamily.multinomial;
+      parms._keep_cross_validation_predictions=false;
+      parms._keep_cross_validation_fold_assignment=true;
+      GBM job = new GBM(parms);
+      gbm = job.trainModel().get();
+
+      assertNotNull(gbm._output._cross_validation_fold_assignment_frame_id);
+      cvFoldAssignmentFrame = DKV.getGet(gbm._output._cross_validation_fold_assignment_frame_id);
```

Comment:
A small comment here: you should ideally have a second copy of the "foldId" Vec, this test would pass even if the modelbuilder completely overwrote the fold column.

Reply:
@michalkurka there are so many things that we can check. Name of the test says `checkFoldAssignmentsAreKeptWithoutMakeCopy` so I think that I've checked this case. This case you are talking about should have been tested in the scope of adding fold_assignment functionality.... check that models are using correct folds. Of course it is easy to extend responsibility of my test....but where is the limit then? I prefer to keep the responsibility/scope as narrow as possible.

Reply:
My point was that using `isBitIdentical` could be misleading here because you are running it on the very same frame/vec. I will leave this one up to you. Not a big deal.

Reply:
@michalkurka Thank you for paying attention to this. I knew this. I even have the test where I'm removing train frame and fold assignment vec is being removed as well. I'm just checking here that we are getting actually the SAME vec as initial vec and assuming we don't mess with vec along the way.